extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts

```diff
@@ -17,31 +18,50 @@ export class GeneralMailFormatter {
     const choices = view.sendBtnModule.popover.choices;
     const recipientsEmails = getUniqueRecipientEmails(newMsgData.recipients);
     if (!choices.encrypt && !choices.sign) { // plain
+      view.S.now('send_btn_text').text('Formatting...');
       return { senderKi: undefined, msg: await new PlainMsgMailFormatter(view).sendableMsg(newMsgData) };
     }
-    let signingPrv: Key | undefined;
     if (!choices.encrypt && choices.sign) { // sign only
-      view.S.now('send_btn_text').text('Signing...');
-      const senderKi = await view.storageModule.getKey(newMsgData.from);
-      signingPrv = await view.storageModule.decryptSenderKey(senderKi);
-      if (!signingPrv) {
+      const senderKis = await view.storageModule.getAccountKeys(newMsgData.from);
+      const signingKey = await GeneralMailFormatter.chooseSigningKeyAndDecryptIt(view, senderKis);
+      if (!signingKey) {
         return undefined;
       }
-      return { senderKi, msg: await new SignedMsgMailFormatter(view).sendableMsg(newMsgData, signingPrv) };
+      view.S.now('send_btn_text').text('Signing...');
+      return { senderKi: signingKey!.keyInfo, msg: await new SignedMsgMailFormatter(view).sendableMsg(newMsgData, signingKey!.key) };
     }
     // encrypt (optionally sign)
-    const result = await view.storageModule.collectSingleFamilyKeys(recipientsEmails, newMsgData.from, choices.sign);
-    if (choices.sign && result.senderKi !== undefined) {
-      signingPrv = await view.storageModule.decryptSenderKey(result.senderKi);
-      if (!signingPrv) {
+    const singleFamilyKeys = await view.storageModule.collectSingleFamilyKeys(recipientsEmails, newMsgData.from, choices.sign);
+    if (singleFamilyKeys.emailsWithoutPubkeys.length) {
```

Comment:
we previously collected senderKi for each family.
As you can see in `if (choices.sign && result.senderKi !== undefined)` signing is optional. It was preferred to have encryption keys (of a single family, i.e. S/MIME) for all the recipients, rather than having a signing key of that family.
Now the signing is actually mandatory, so the test crashes (and only the test, but the send button too)

Reply:
It's a good idea to pick the signing key in `collectSingleFamilyKeys`. We can adjust the sorting function that picks a family.

Reply:
Thanks! This was the method I was worried about the most. I'll have a look.